### ISSUE TITLE
Small source fix makes using BoringSSL easier

### ISSUE THREAD

**Issue by byllyfish** [chriskohlhoff/asio#52]
I'm using BoringSSL with ASIO 1.11. BoringSSL is Google's downstream fork of OpenSSL that removes some of the cruft from OpenSSL while remaining source compatible. I've found three issues with ASIO, but I only really need one source fix (which is openssl-fork-agnostic):

```diff
diff --git a/asio/include/asio/ssl/detail/impl/openssl_init.ipp b/asio/include/asio/ssl/detail/impl/openssl_init.ipp
index 2c40d40..0229374 100644
--- a/asio/include/asio/ssl/detail/impl/openssl_init.ipp
+++ b/asio/include/asio/ssl/detail/impl/openssl_init.ipp
@@ -63,7 +63,11 @@ public:
     ::CRYPTO_set_id_callback(0);
     ::CRYPTO_set_locking_callback(0);
     ::ERR_free_strings();
+#if OPENSSL_VERSION_NUMBER >= 0x10000000L
+    ::ERR_remove_thread_state(NULL);
+#else
     ::ERR_remove_state(0);
+#endif // OPENSSL_VERSION_NUMBER >= 0x10000000L
     ::EVP_cleanup();
     ::CRYPTO_cleanup_all_ex_data();
     ::CONF_modules_unload(1);
```

ERR_remove_state was deprecated in favor of ERR_remove_thread_state in OpenSSL 1.0.0 (although it remained for binary compatibility).

https://www.openssl.org/docs/crypto/ERR_remove_state.html

The other two issues are:

1. `CONF_modules_unload` isn't declared or defined in BoringSSL. Config modules aren't supported.
2. `SSL_R_SHORT_READ` isn't defined in BoringSSL. Unused error codes have been removed.

These last two issues can be handled in my own code with a header prefix:

```c++
#if defined(OPENSSL_IS_BORINGSSL)
extern "C" {
#if !defined(SSL_R_SHORT_READ)
# define SSL_R_SHORT_READ    SSL_R_UNEXPECTED_RECORD
#endif // !defined(SSL_R_SHORT_READ)
inline void CONF_modules_unload(int p) {}
}
#endif // defined(OPENSSL_IS_BORINGSSL)
```

**Comment by thughes** [comment]
+1

FWIW, it looks like `SSL_R_SHORT_READ` has been removed in openssl as well: https://github.com/openssl/openssl/commit/45f55f6a5bdcec411ef08a6f8aae41d5d3d234ad

**Comment by chriskohlhoff** [comment]
Fixed in commits 628e3ca9fe7a1bed1ce2308e2df4a1a4ecd1dfe7, 5fa80539834c10406611bb02c20cdba2a9171f4a and 92bfc623e6a71353dd2c783f4c9fef5591ac550d.

Note that a new `ssl::error::stream_truncated` error code has been added, rather than reusing `SSL_R_SHORT_READ`.

**Comment by RajatKumar007** [comment]
I am trying to integrate boringssl with boost asio version 1.10.7. But getting several compilation error as :

boost/asio/ssl/impl/context.ipp:208:16: error: member access into incomplete type 'SSL_CTX' (aka 'ssl_ctx_st')
    if (handle_->default_passwd_callback_userdata)
               ^
boringssl/include\openssl/base.h:333:16: note: forward declaration of 'ssl_ctx_st'
typedef struct ssl_ctx_st SSL_CTX;


boost/asio/ssl/detail/impl/engine.ipp:207:34: error: too many arguments provided to function-like macro invocation
          ERR_PACK(ERR_LIB_SSL, 0, SSL_R_SHORT_READ),
                                   ^
boringssl/include\openssl/err.h:446:9: note: macro 'ERR_PACK' defined here
  #define ERR_PACK(lib, reason)
